File: Java/백준/Silver/2564. 경비원/경비원.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws Exception {
        Scanner sc = new Scanner(System.in);
        int W = sc.nextInt();  // 가로 길이
        int H = sc.nextInt();  // 세로 길이
        
        int C = sc.nextInt();  // 상점의 개수
        
        int[][] arr = new int[C][2];
        
        // 상점 위치 입력 받기
        for (int c = 0; c < C; c++) {
            arr[c][0] = sc.nextInt();
            arr[c][1] = sc.nextInt();
        }
        
        // 동근이 위치 입력 받기
        int sDir = sc.nextInt();
        int sP = sc.nextInt();
        
        int answer = 0;
        
        // 상점마다 거리를 계산
        for (int i = 0; i < C; i++) {
            if (sDir == arr[i][0]) {
                answer += Math.abs(arr[i][1] - sP);
            } else if (sDir == 1) {
                if (arr[i][0] == 2) {
                    answer += (H + Math.min((sP + arr[i][1]), ((W-sP) + (W-arr[i][1])))); 
                } else if (arr[i][0] == 3) {
                    answer += (sP + arr[i][1]);
                } else if (arr[i][0] == 4) {
                    answer += ((W-sP) + arr[i][1]);
                }
            } else if (sDir == 2) {
                if (arr[i][0] == 1) {
                    answer += (H + Math.min((sP + arr[i][1]), ((W-sP) + (W-arr[i][1])))); 
                } else if (arr[i][0] == 3) {
                    answer += (sP + (H - arr[i][1]));
                } else if (arr[i][0] == 4) {
                    answer += ((W-sP) + (H - arr[i][1]));
                }
            } else if (sDir == 3) {
                if (arr[i][0] == 1) {
                    answer += sP + arr[i][1];
                } else if (arr[i][0] == 2) {
                    answer += (H-sP) + arr[i][1];
                } else if (arr[i][0] == 4) {
                    answer += (W + Math.min((sP + arr[i][1]), ((H-sP) + (H-arr[i][1]))));
                }
            } else if (sDir == 4) {
                if (arr[i][0] == 1) {
                    answer += sP + (W - arr[i][1]);
                } else if (arr[i][0] == 2) {
                    answer += (H-sP) + (W - arr[i][1]);
                } else if (arr[i][0] == 3) {
                    answer += (W + Math.min((sP + arr[i][1]), ((H-sP) + (H-arr[i][1]))));
                }
            }
        }
        
        System.out.println(answer);
    }
}

File: node.js/백준/Silver/2564. 경비원/경비원.js
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n");

// dir : 1 - 북, 2 - 남, 3 - 서, 4 - 동

let idx = 0;

const [W, H] = input[idx].split(" ").map(Number);
idx++;

const T = parseInt(input[idx]);
idx++;

const stores = [];

for (let t = 0; t < T; t++) {
  const [dir, dist] = input[idx].split(" ").map(Number);
  stores.push([dir, dist]);
  idx++;
}

const [sDir, sDist] = input[idx].split(" ").map(Number);

/**
 * 같은 방위에 있을 때,
 * 오른쪽에 있을 때, 1 -> 3, 2 -> 3, 3 -> 1, 3 -> 2,
 * 왼쪽에 있을 때,
 * 맞은 편에 있을 때,
 */

let answer = 0;

for (let [tDir, tDist] of stores) {
  if (sDir == tDir) {
    answer += Math.abs(sDist - tDist);
    continue;
  }

  if (tDir == 1) {
    if (sDir == 2) {
      answer += Math.min(H + tDist + sDist, H + W - tDist + W - sDist);
      continue;
    } else if (sDir == 3) {
      answer += tDist + sDist;
      continue;
    } else if (sDir == 4) {
      answer += W - tDist + sDist;
      continue;
    }
  }

  if (tDir == 2) {
    if (sDir == 1) {
      answer += Math.min(H + tDist + sDist, H + W - tDist + W - sDist);
      continue;
    } else if (sDir == 3) {
      answer += tDist + H - sDist;
      continue;
    } else if (sDir == 4) {
      answer += W - tDist + H - sDist;
      continue;
    }
  }

  if (tDir == 3) {
    if (sDir == 1) {
      answer += sDist + tDist;
      continue;
    } else if (sDir == 2) {
      answer += sDist + H - tDist;
      continue;
    } else if (sDir == 4) {
      answer += Math.min(W + sDist + tDist, W + H - sDist + H - tDist);
      continue;
    }
  }

  if (tDir == 4) {
    if (sDir == 1) {
      answer += W - sDist + tDist;
      continue;
    } else if (sDir == 2) {
      answer += W - sDist + H - tDist;
      continue;
    } else if (sDir == 3) {
      answer += Math.min(W + sDist + tDist, W + H - sDist + H - tDist);
    }
  }
}

console.log(answer);
